### ISSUE TITLE
Implicit-explicit Schwarz coupling is broken

### ISSUE THREAD

**Issue by ikalash** [sandialabs/LCM#51]
I discovered yesterday that implicit-explicit Schwarz coupling is broken in LCM.  It slipped through the cracks due to us not having a test of the capability.  One of the issues was in Piro, which I have fixed in this PR: https://github.com/trilinos/Trilinos/pull/9935 .  There is still a problem, it looks like in StrongSchwarzBC_Def.hpp when calling getNumEquations on this_app.  Below is the full backtrace of the error.  @lxmota I might ask you to look at this since you put in the this_app/coupled_app logic and relevant routines.  

```
(gdb) run clamped_impl-expl.yaml 
Starting program: /home/ikalash/ACE/Albany-LCM-master/build-dbg/tests/LCM/Schwarz/Alternating/Dynamics/Clamped/Albany clamped_impl-expl.yaml
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
[Detaching after fork from child process 3062340]
[New Thread 0x7fffd7b7c640 (LWP 3062344)]
[New Thread 0x7fffd71b7640 (LWP 3062345)]
[New Thread 0x7fffd5908640 (LWP 3062350)]

   ###    ##       ########     ###    ##    ## ##    ##
  ## ##   ##       ##     ##   ## ##   ###   ##  ##  ## 
 ##   ##  ##       ##     ##  ##   ##  ####  ##   ####  
##     ## ##       ########  ##     ## ## ## ##    ##   
######### ##       ##     ## ######### ##  ####    ##   
##     ## ##       ##     ## ##     ## ##   ###    ##   
##     ## ######## ########  ##     ## ##    ##    ##   

** Trilinos git commit id - 37e9418d718
** Albany git branch ------ master
** Albany git commit id --- 37bc00f41
** Albany cxx compiler ---- GNU 11.2.1
** Simulation start time -- 2021-11-16 at 10:52:19
***************************************************************
Problem Name = Mechanics 3D
Solution Method = Transient Tempus
Mechanics problem:
	Spatial dimension             : 3
	Mechanics variables           : DOF
	Temperature variables         : None
	Dislocation Density variables : None
	Pore Pressure variables       : None
	Transport variables           : None
	HydroStress variables         : None
	Damage variables              : None
	Stabilized Pressure variables : None
Initializing material database from material_00.yaml
TmplSTKMeshStruct:: Creating 3D mesh of size 1x1x75 elements and scaled to 0.01x0.01x0.75
Num MeshSpecs: 1
Calling MechanicsProblem::buildEvaluators
Calling MechanicsProblem::constructDirichletEvaluators
Calling MechanicsProblem::constructNeumannEvaluators
[GenericSTKMeshStruct] Processing field requirements...
STKDisc: 75 elements on Proc 0 
STKDisc: nodeset NodeSet0 has size 152  on Proc 0.
STKDisc: nodeset NodeSet1 has size 152  on Proc 0.
STKDisc: nodeset NodeSet2 has size 152  on Proc 0.
STKDisc: nodeset NodeSet3 has size 152  on Proc 0.
STKDisc: nodeset NodeSet4 has size 4  on Proc 0.
STKDisc: nodeset NodeSet5 has size 4  on Proc 0.
STKDisc: nodeset NodeSet99 has size 1  on Proc 0.
STKDisc: sideset SideSet0 has size 75  on Proc 0.
STKDisc: sideset SideSet1 has size 75  on Proc 0.
STKDisc: sideset SideSet2 has size 75  on Proc 0.
STKDisc: sideset SideSet3 has size 75  on Proc 0.
STKDisc: sideset SideSet4 has size 1  on Proc 0.
STKDisc: sideset SideSet5 has size 1  on Proc 0.
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 Sacado ParameterLibrary has been initialized:
 Library of all registered parameters:
	SDBC on NS NodeSet0 for DOF X:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet1 for DOF X:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet2 for DOF Y:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet3 for DOF Y:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet4 for DOF Z:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet5 for DOF StrongSchwarz:  Supports AD = 1, Supports_Analytic = 0
	Time:  Supports AD = 1, Supports_Analytic = 0
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Number of parameter vectors  = 0
Number of response vectors  = 1
Number of distributed parameters vectors  = 0

A) Get the base parameter list ...

    Using Tempus 'Time Step Control'.

C) Create and initalize the forward model ...

D) Create the stepper and integrator for the forward problem ...
 Warning -- Did not find a Tempus Time Events to create!
 Should call add() with this (app-specific?) Time Event.
 
Problem Name = Mechanics 3D
Solution Method = Transient Tempus
Mechanics problem:
	Spatial dimension             : 3
	Mechanics variables           : DOF
	Temperature variables         : None
	Dislocation Density variables : None
	Pore Pressure variables       : None
	Transport variables           : None
	HydroStress variables         : None
	Damage variables              : None
	Stabilized Pressure variables : None
Initializing material database from material_01.yaml
TmplSTKMeshStruct:: Creating 3D mesh of size 1x1x75 elements and scaled to 0.01x0.01x0.75
Num MeshSpecs: 1
Calling MechanicsProblem::buildEvaluators
Calling MechanicsProblem::constructDirichletEvaluators
Calling MechanicsProblem::constructNeumannEvaluators
[GenericSTKMeshStruct] Processing field requirements...
STKDisc: 75 elements on Proc 0 
STKDisc: nodeset NodeSet0 has size 152  on Proc 0.
STKDisc: nodeset NodeSet1 has size 152  on Proc 0.
STKDisc: nodeset NodeSet2 has size 152  on Proc 0.
STKDisc: nodeset NodeSet3 has size 152  on Proc 0.
STKDisc: nodeset NodeSet4 has size 4  on Proc 0.
STKDisc: nodeset NodeSet5 has size 4  on Proc 0.
STKDisc: nodeset NodeSet99 has size 1  on Proc 0.
STKDisc: sideset SideSet0 has size 75  on Proc 0.
STKDisc: sideset SideSet1 has size 75  on Proc 0.
STKDisc: sideset SideSet2 has size 75  on Proc 0.
STKDisc: sideset SideSet3 has size 75  on Proc 0.
STKDisc: sideset SideSet4 has size 1  on Proc 0.
STKDisc: sideset SideSet5 has size 1  on Proc 0.
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
 Sacado ParameterLibrary has been initialized:
 Library of all registered parameters:
	SDBC on NS NodeSet0 for DOF X:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet1 for DOF X:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet2 for DOF Y:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet3 for DOF Y:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet4 for DOF StrongSchwarz:  Supports AD = 1, Supports_Analytic = 0
	SDBC on NS NodeSet5 for DOF Z:  Supports AD = 1, Supports_Analytic = 0
	Time:  Supports AD = 1, Supports_Analytic = 0
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

Number of parameter vectors  = 0
Number of response vectors  = 1
Number of distributed parameters vectors  = 0

A) Get the base parameter list ...

    Using Tempus 'Time Step Control'.

C) Create and initalize the forward model ...

D) Create the stepper and integrator for the forward problem ...
 Warning -- Did not find a Tempus Time Events to create!
 Should call add() with this (app-specific?) Time Event.
 

Thread 1 "Albany" received signal SIGSEGV, Segmentation fault.
0x00007ffff6f9e89c in Albany::Application::getNumEquations (this=0x0)
    at /home/ikalash/ACE/Albany-LCM-master/src/Albany_Application.hpp:157
157	    return neq;
Missing separate debuginfos, use: dnf debuginfo-install blas-3.9.0-6.fc34.x86_64 cyrus-sasl-lib-2.1.27-8.fc34.x86_64 glibc-2.33-20.fc34.x86_64 hdf-4.2.15-4.fc34.x86_64 hdf-libs-4.2.15-4.fc34.x86_64 hdf5-1.10.6-5.fc34.x86_64 hwloc-libs-2.4.1-3.fc34.x86_64 infinipath-psm-3.3-26_g604758e_open.6.fc34.3.x86_64 keyutils-libs-1.6.1-2.fc34.x86_64 krb5-libs-1.19.2-2.fc34.x86_64 lapack-3.9.0-6.fc34.x86_64 libaec-1.0.5-1.fc34.x86_64 libbrotli-1.0.9-4.fc34.x86_64 libcurl-7.76.1-7.fc34.x86_64 libevent-2.1.12-3.fc34.x86_64 libfabric-1.12.1-1.fc34.x86_64 libgcc-11.2.1-1.fc34.x86_64 libgfortran-11.2.1-1.fc34.x86_64 libibverbs-36.0-1.fc34.x86_64 libnghttp2-1.43.0-2.fc34.x86_64 libpsl-0.21.1-3.fc34.x86_64 libquadmath-11.2.1-1.fc34.x86_64 librdmacm-36.0-1.fc34.x86_64 libssh-0.9.5-2.fc34.x86_64 libstdc++-11.2.1-1.fc34.x86_64 libtirpc-1.3.2-0.fc34.x86_64 libunistring-0.9.10-10.fc34.x86_64 libxcrypt-4.4.25-1.fc34.x86_64 netcdf-4.7.3-6.fc34.x86_64 numactl-libs-2.0.14-3.fc34.x86_64 openldap-2.4.57-5.fc34.x86_64 openmpi-4.1.0-5.fc34.x86_64 openssl-libs-1.1.1l-1.fc34.x86_64 pcre2-10.36-4.fc34.x86_64 pmix-3.2.3-1.fc34.x86_64 ucx-1.10.1-1.fc34.x86_64 zlib-1.2.11-26.fc34.x86_64
(gdb) bt
#0  0x00007ffff6f9e89c in Albany::Application::getNumEquations (this=0x0)
    at /home/ikalash/ACE/Albany-LCM-master/src/Albany_Application.hpp:157
#1  0x00007ffff2ef0d22 in LCM::StrongSchwarzBC_Base<PHAL::AlbanyTraits::Residual, PHAL::AlbanyTraits>::computeBCsDTK (
    this=0xbd8c00) at /home/ikalash/ACE/Albany-LCM-master/src/LCM/evaluators/bc/StrongSchwarzBC_Def.hpp:280
#2  0x00007ffff2eef125 in LCM::fillSolution<LCM::StrongSchwarzBC<PHAL::AlbanyTraits::Residual, PHAL::AlbanyTraits>, PHAL::AlbanyTraits> (sbc=..., workset=...) at /home/ikalash/ACE/Albany-LCM-master/src/LCM/evaluators/bc/StrongSchwarzBC_Def.hpp:468
#3  0x00007ffff2eee069 in LCM::StrongSchwarzBC<PHAL::AlbanyTraits::Residual, PHAL::AlbanyTraits>::preEvaluate (
    this=0xbd8c00, workset=...) at /home/ikalash/ACE/Albany-LCM-master/src/LCM/evaluators/bc/StrongSchwarzBC_Def.hpp:612
#4  0x00007ffff6f36f72 in PHX::DagManager<PHAL::AlbanyTraits>::preEvaluate (this=0xba94e8, d=...)
    at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Phalanx_DAG_Manager_Def.hpp:641
#5  0x00007ffff6f33a7f in PHX::EvaluationContainer<PHAL::AlbanyTraits::Residual, PHAL::AlbanyTraits>::preEvaluate (
    this=0xba94e0, d=...) at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Phalanx_EvaluationContainer_Def.hpp:502
#6  0x00007ffff6ec3612 in PHX::FieldManager<PHAL::AlbanyTraits>::preEvaluate<PHAL::AlbanyTraits::Residual> (this=0xbe0e60, 
    d=...) at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Phalanx_FieldManager_Def.hpp:349
#7  0x00007ffff6eaf3ec in Albany::Application::computeGlobalResidualImpl (this=0xb3bd20, current_time=0, x=..., x_dot=..., 
    x_dotdot=..., p=..., f=..., dt=0) at /home/ikalash/ACE/Albany-LCM-master/src/Albany_Application.cpp:1142
#8  0x00007ffff6eb03c7 in Albany::Application::computeGlobalResidual (this=0xb3bd20, current_time=0, x=..., x_dot=..., 
    x_dotdot=..., p=..., f=..., dt=0) at /home/ikalash/ACE/Albany-LCM-master/src/Albany_Application.cpp:1227
#9  0x00007ffff6f55a84 in Albany::ModelEvaluator::evalModelImpl (this=0xf91190, inArgs=..., outArgs=...)
    at /home/ikalash/ACE/Albany-LCM-master/src/Albany_ModelEvaluator.cpp:679
#10 0x00007ffff6b69809 in Thyra::ModelEvaluatorDefaultBase<double>::evalModel (this=0xf91190, inArgs=..., outArgs=...)
    at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Thyra_ModelEvaluatorDefaultBase.hpp:709
#11 0x00007ffff6b80251 in Thyra::DefaultModelEvaluatorWithSolveFactory<double>::evalModelImpl (this=0xebbc90, inArgs=..., 
    outArgs=...) at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Thyra_DefaultModelEvaluatorWithSolveFactory.hpp:325
#12 0x00007ffff6b69809 in Thyra::ModelEvaluatorDefaultBase<double>::evalModel (this=0xebbcd8, inArgs=..., outArgs=...)
    at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Thyra_ModelEvaluatorDefaultBase.hpp:709
#13 0x00007ffff6b79e3d in Piro::InvertMassMatrixDecorator<double>::evalModelImpl (this=0xec2910, inArgs=..., outArgs=...)
    at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Piro_InvertMassMatrixDecorator_Def.hpp:348
#14 0x00007ffff6b69809 in Thyra::ModelEvaluatorDefaultBase<double>::evalModel (this=0xec2910, inArgs=..., outArgs=...)
    at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Thyra_ModelEvaluatorDefaultBase.hpp:709
#15 0x00007fffefc01caf in Tempus::StepperExplicit<double>::evaluateExplicitODE (this=0xf2c850, xDotDot=..., x=..., xDot=..., 
    time=0, p=...) at /home/ikalash/ACE/Trilinos/packages/tempus/src/Tempus_StepperExplicit_impl.hpp:313
#16 0x00007fffefcb1d32 in Tempus::StepperNewmarkExplicitAForm<double>::setInitialConditions (this=0xf2c820, 
    solutionHistory=...) at /home/ikalash/ACE/Trilinos/packages/tempus/src/Tempus_StepperNewmarkExplicitAForm_impl.hpp:187
#17 0x00007fffefb3acf8 in Tempus::IntegratorBasicOld<double>::initializeSolutionHistory (this=0xe1d2c0, state=...)
    at /home/ikalash/ACE/Trilinos/packages/tempus/src/Tempus_IntegratorBasicOld_impl.hpp:149
--Type <RET> for more, q to quit, c to continue without paging--
#18 0x00007fffefb3b563 in Tempus::IntegratorBasicOld<double>::setSolutionHistory (this=0xe1d2c0, sh=...)
    at /home/ikalash/ACE/Trilinos/packages/tempus/src/Tempus_IntegratorBasicOld_impl.hpp:208
#19 0x00007fffefb3c44f in Tempus::IntegratorBasicOld<double>::initialize (this=0xe1d2c0)
    at /home/ikalash/ACE/Trilinos/packages/tempus/src/Tempus_IntegratorBasicOld_impl.hpp:292
#20 0x00007fffefb36386 in Tempus::IntegratorBasicOld<double>::IntegratorBasicOld (this=0xe1d2c0, inputPL=..., model=..., 
    __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /home/ikalash/ACE/Trilinos/packages/tempus/src/Tempus_IntegratorBasicOld_impl.hpp:28
#21 0x00007fffefb3eef3 in Tempus::integratorBasic<double> (pList=..., model=...)
    at /home/ikalash/ACE/Trilinos/packages/tempus/src/Tempus_IntegratorBasicOld_impl.hpp:667
#22 0x00007ffff6b38522 in Piro::TempusIntegrator<double>::TempusIntegrator (this=0xf2c7b0, pList=..., model=..., 
    sens_method=Piro::NONE) at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Piro_TempusIntegrator_Def.hpp:63
#23 0x00007ffff6b29cc6 in Piro::TempusSolver<double>::initialize (this=0xe57d60, appParams=..., in_model=..., 
    in_adjointModel=...) at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Piro_TempusSolver_Def.hpp:342
#24 0x00007ffff6b18bb3 in Piro::TempusSolver<double>::TempusSolver (this=0xe57d60, appParams=..., in_model=..., 
    in_adjointModel=..., piroObserver=..., __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Piro_TempusSolver_Def.hpp:120
#25 0x00007ffff6b0af29 in Piro::tempusSolver<double> (appParams=..., in_model=..., in_adjointModel=..., piroObserver=...)
    at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Piro_TempusSolver_Def.hpp:755
#26 0x00007ffff6afe5a1 in Piro::SolverFactory::createSolver<double> (this=0x7fffffff9d7f, piroParams=..., model=..., 
    adjointModel=..., observer=...) at /home/ikalash/ACE/Trilinos/build-dbg/install/include/Piro_SolverFactory_Def.hpp:159
#27 0x00007ffff6ad41ae in Albany::SolverFactory::createAndGetAlbanyApp (this=0x7fffffffa520, albanyApp=..., appComm=..., 
    solverComm=..., initial_guess=..., createAlbanyApp=true)
    at /home/ikalash/ACE/Albany-LCM-master/src/Albany_SolverFactory.cpp:231
#28 0x00007ffff2bca198 in LCM::SchwarzAlternating::SchwarzAlternating (this=0x78bfb0, app_params=..., comm=..., 
    __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at /home/ikalash/ACE/Albany-LCM-master/src/LCM/solvers/Schwarz_Alternating.cpp:196
#29 0x00007ffff6ad3869 in Albany::SolverFactory::createAndGetAlbanyApp (this=0x7fffffffbbd0, albanyApp=..., appComm=..., 
    solverComm=..., initial_guess=..., createAlbanyApp=true)
    at /home/ikalash/ACE/Albany-LCM-master/src/Albany_SolverFactory.cpp:184
#30 0x0000000000466d48 in main (argc=2, argv=0x7fffffffd048) at /home/ikalash/ACE/Albany-LCM-master/src/Main_Solve.cpp:101
```

**Comment by ikalash** [comment]
I forgot to mention: to reproduce the issue, please run: https://github.com/sandialabs/LCM/blob/master/tests/LCM/Schwarz/Alternating/Dynamics/Clamped/clamped_impl-expl.yaml .  I added this test today and it will fail in the nightlies starting tonight.

**Comment by ikalash** [comment]
Also, if not clear above, debugging this issue in Albany requires using this fork of Trilinos for the time being (until it gets merged in): https://github.com/trilinos/Trilinos/tree/piro-fix-for-lcm .

**Comment by ikalash** [comment]
Here is the failure on CDash: https://sems-cdash-son.sandia.gov/cdash/test/1851682  .

**Comment by ikalash** [comment]
Some additional information...

The problem is when we have an explicit scheme involved in the coupling.  The seg fault always happens in the explicit domain.  It appears for the explicit stepper, StrongSchwarzBC gets called in the initial step, which does not happen for implicit.  Perhaps this is due to some change in how Tempus handles the initial step with explicit time-stepping?  So implicit-explicit coupling per se is not the problem.

Code throws here https://github.com/sandialabs/LCM/blob/master/src/LCM/evaluators/bc/StrongSchwarzBC_Def.hpp#L280 b/c ```this_app.getNumEquations()``` gives an address not mapped.  Not sure why yet.

**Comment by lxmota** [comment]
I wrote this check:

https://github.com/sandialabs/LCM/blob/master/src/LCM/evaluators/bc/StrongSchwarzBC_Def.hpp#L280

And I remember that the reason I put it there is to make sure that we were coupling the same physics. But it's not really necessary for LCM since we are dealing with solid mechanics only. Maybe it can be removed and the issue of the segfault will be resolved, if that is the only problem.

**Comment by ikalash** [comment]
So, I know what is going on, sort of, but I'm not sure exactly why it is a problem and how to fix it.  It looks like some time back (starting in October 2018), the way the initial condition for the acceleration in the Tempus explicit Newmark stepper was changed.  There is all this code now allowing one to do different "consistency checks" for the initial condition: https://github.com/trilinos/Trilinos/blob/master/packages/tempus/Tempus_StepperNewmarkExplicitAForm_impl.hpp#L158-L198 .  It looks like this is where you actually tell the code how to calculate the initial acceleration.  What we want is the "Consistent" option, I think.  However, if that's selected (the default), the code tries to call StrongSchwarzBC in the initial step, and seg faults b/c I guess the number of equations for the present app has not been set yet (I'm not sure why).  If I set "Initial Condition Consistency" = "None" (or something other than "Consistent"), the code runs but the BCs don't propagate correctly to the explicit domain in the Schwarz method.  @lxmota any thoughts?

Also, I don't think just removing the assert will help.  If you look at line 283 in StrongSchwarzBC_Def.hpp, it tries to get neq from this_app.  That's the thing that is undefined, so the code would still seg fault.

What's weird is I'm sure we ran Schwarz after 10/2018 with implicit-explicit coupling.  10/2018 is then the initial condition consistency check was first put in.  There must have been something added later that actually broke things.  I guess I could try to build some older version of the codes to figure it out.  It's really unfortunate we didn't have an implicit-explicit or explicit-explicit test for Schwarz in the test harness...

**Comment by ikalash** [comment]
Tagging @ccober6 - we might bug you about this problem if we can't figure it out...

**Comment by ccober6** [comment]
So the "Consistent" option is what I think you want.  "Consistent" means for the initial conditions (t=0) that x, xDot, and xDotDot satisfy xDotDot = f(xDot, x, t=0) for explicit problems, and f(xDotDot, xDot, x, t=0) = 0 for implicit problems.

For explicit problems, it calls f(xDot, x, t=0) ModelEvaluator and sets xDotDot.

For implicit problems, it solves the nonlinear (usually) problem f(xDotDot, xDot, x, t=0) = 0 for xDotDot.

So it sounds like the stepper->setInitialConditions() is being called before everything is setup?  @ikalash, I think we had an issue a few months? back where the Tempus setup order in Albany needed to be changed.  Could that be related?

Also, we had a "weird behavior" from August 10, 2020, related to FSAL and setting BCs
[Re Tempus weird behavior  a couple questions.pdf](https://github.com/sandialabs/LCM/files/7572365/Re.Tempus.weird.behavior.a.couple.questions.pdf)
Not sure if it factors in.

Also I am not clear about the implicit-explicit and explicit-explicit part of the discussion.  Are you doing operator splitting?  Do you have two ModelEvaluators?

**Comment by ikalash** [comment]
@ccober6 : thanks for the reply and for trying to help.  Yes, it does sound like stepper->setInitialConditions() is being called before everything is setup.  Did the time at which stepper->setInitialConditions() gets called change in Tempus at any point in the recent future?  

Is this the recent issue you're referring to: https://github.com/trilinos/Trilinos/issues/9767 ?  I didn't really think this issue was related.  I did have some suspicious about Use FSAL or related code.  No matter what I set "Use FSAL" to, I still get the error.

I'm currently trying to get a build from earlier in which the case ran.  That will give me an idea of when the problem started at least.  We submitted our paper on this originally in September 2020, so the runs were made before then, and the case was working.

The implicit-explicit and explicit-explicit part refers to our Schwarz coupling.  I think you're at least somewhat familiar with how we are doing alternating Schwarz-based coupling in Albany.  We solve a sequence of problems in two overlapping subdomains in an alternating iterative fashion.  The framework lets us use an explicit stepper in one subdomain and an implicit one in another - this is what implicit-explicit refers to.  What I discovered is if you do any kind of coupling using explicit, this problem shows up.  In contrast, just running a problem with an explicit Newmark stepper w/o Schwarz coupling works fine in Albany.  Albany has a special model evaluator that sets up all the solvers for the coupling, so something there is not playing nicely with the way Tempus works currently.

**Comment by ccober6** [comment]
@ikalash : it could be related to changes with https://github.com/trilinos/Trilinos/pull/8905 (~April 2021).  There was some cleanup of IntegratorBasic::initializeSolutionHistory(), but I do not think there was an order change, but I could be wrong.  :)

**Comment by ikalash** [comment]
So I have some additional information, but unfortunately I'm pretty much stuck.  Here is some additional information:

- The backtrace / seg fault reported above happens only with DTK.  W/o DTK, implicit-explicit runs, but does not give the correct solution with the current version of the code.
- The incorrect solution issue for implicit-explicit coupling started sometime b/w Aug. 1 2020 and Jan. 1 2021.  I can probably figure out what commit specifically broke the case if we wanted to.
-  I pulled Trilinos and Albany from a number of different times dating back to 2018 to try to find what Trilinos commit broke things.  A seg fault for the couplings with the first subdomain being explicit happens in all the builds.  This is truly bizarre in my opinion b/c explicit-implicit and explicit-explicit definitely worked in 2018-2019, when I made the runs for our paper.

Meeting tomorrow with @lxmota to discuss how we want to proceed.

**Comment by ikalash** [comment]
Actually I have a lead: looks like things worked Aug. 1 2018.  Worst case scenario we can give the sha corresponding to this date in our paper.  Then we'll want to release the input files in a separate repo I think, rather than as part of Albany.  Something broke b/w Aug. 1 2018 - Aug. 1 2019.  I can try to figure out what it was.  I can't believe it was so long ago...

**Comment by ikalash** [comment]
I've tracked this down more and it looks like Schwarz coupling with explicit steppers got broken due to changes in Trilinos several times.  The first time was in late March 2019.  At this time, Schwarz broke for explicit-explicit and explicit-implicit coupling (coupling where explicit is first).  Later, implicit-explicit coupling got broken too, which I'm still trying to track down.  

The first problem appears to be due to changes in setInitialState in March 2019.  If I take the following sha of Trilinos c9264fa372b5366e8f8fd6c914a1938d199a6df8 and revert the following two commits, the issue goes away: 

```
commit 9592b565f29f48a144a470f388f346fb334a8238
Author: Irina K. Tezaur <ikalash@sandia.gov>
Date:   Sun Nov 28 19:08:48 2021 -0700

    Revert "Merge pull request #4491 from trilinos/tempus-add-stepper-initial-conditions-3715"
    
    This reverts commit 9d4360d65fb90e33863b32a540175053e6cb82b0, reversing
    changes made to 6b69e56d91373a7e2f6a7112c10a33718f5eb9a6.
    
    Conflicts:
        packages/tempus/src/Tempus_StepperExplicit_decl.hpp
        packages/tempus/src/Tempus_StepperExplicit_impl.hpp
        packages/tempus/src/Tempus_StepperHHTAlpha_decl.hpp
        packages/tempus/src/Tempus_StepperImplicit_impl.hpp
        packages/tempus/src/Tempus_StepperNewmarkExplicitAForm_impl.hpp
        packages/tempus/src/Tempus_StepperNewmarkImplicitAForm_impl.hpp
        packages/tempus/src/Tempus_StepperNewmarkImplicitDForm_decl.hpp
        packages/tempus/src/Tempus_StepperOperatorSplit_decl.hpp
        packages/tempus/src/Tempus_StepperStaggeredForwardSensitivity_decl.hpp
        packages/tempus/src/Tempus_WrapperModelEvaluatorPairIMEX_Basic_decl.hpp
        packages/tempus/src/Tempus_WrapperModelEvaluatorPairPartIMEX_Basic_decl.hpp
        packages/tempus/src/Tempus_WrapperModelEvaluatorSecondOrder_decl.hpp

commit d6b6965c4ee4484e8dec343c5ad2f48014a56162
Author: Irina K. Tezaur <ikalash@sandia.gov>
Date:   Sun Nov 28 15:24:58 2021 -0700

    Revert "Piro: fixing Albany compilation error by replacing setInitialState()"
    
    This reverts commit 02b1c5cd2a26f8867f12d449813e8f3b2c1bdd50.
```
I have a build with the revert here on cee-compute021: /scratch/ikalash/Trilinos-Apr1-2019-with-revert.  This Trilinos works with the following Albany: b9a0d993f0cb31417a0c0adf86048616a3a652e1 . 

@ccober: could it be that the switch from setInitialState to setInitialSolutionHistory (https://github.com/trilinos/Trilinos/commit/02b1c5cd2a26f8867f12d449813e8f3b2c1bdd50) changed how things are being initialized for explicit Newmark?  Unless there is a simple fix, we don't have the resources to fix this right now unfortunately, as Schwarz development in Albany is no longer funded.  I am more optimistic about debugging implicit-explicit, which broke much later and for which I have some leads.  Will post here when I have more information.

For documentation, here is the backtrace from gdb for the explicit-implicit and explicit-explicit coupling error: [gdb_out.txt](https://github.com/sandialabs/LCM/files/7615301/gdb_out.txt)

**Comment by ccober6** [comment]
@ikalash: So the switch from setInitialState to setInitialSolutionHistory (trilinos/Trilinos@02b1c5c) was to incorporate the changes of (https://github.com/trilinos/Trilinos/pull/4491/files#diff-0c811b7e336240dad1f1662836cfe716b886a5b4bbe7647f41a64ccefb94d364).  See packages/tempus/src/Tempus_IntegratorBasic_impl.hpp:166.  The two changes are the name of the function (setInitialSolutionHistory) and adding a call to stepper_->setInitialConditions(solutionHistory_);

This commit added the new "Initial Condition Consistency" capability.  If it is set to 'None' you should get the previous behavior, as it should do nothing as before (i.e., it was assumed that the ICs were set by the application).  But this is probably the additional call you might be looking for, as you can set the initial conditions at this point.  If the ModelEvaluator is not completely setup, then you could see an error from this call when you use 'Consistent' or 'App'.

So at this point in the commit history, a new call was added to stepper_->setInitialConditions(solutionHistory_) to set the ICs, and not a change in the order of calls.

So it looks like the Schwarz coupling does not work with 'Consistent' ICs.  A work around would be to set "Initial Condition Consistency" to 'None', finish the Albany (ModelEvaluator) setup, and before starting the advanceTime()

- stepper->setICConsistency("Consistent")
- stepper_->setInitialConditions(solutionHistory_);

or the fix would be to finish the Albany (ModelEvaluator) setup and then finish the IntegratorBasic setup?

Hope this helps.

**Comment by lxmota** [comment]
Closing as won't fix — irrelevant after recent code changes.
